Mathlib/Combinatorics/SimpleGraph/Diameter.lean

```diff
@@ -0,0 +1,135 @@
+in the graph.
+
+## Main definitions
+
+- `SimpleGraph.diam` is the graph diameter.
+
+- `SimpleGraph.ediam` is the graph extended diameter.
+
+## Todo
+
+- Prove that `ENat.toNat G.girth ≤ 2 * G.diam + 1` when the diameter is non-zero.
+
+-/
+
+namespace SimpleGraph
+variable {α : Type*} {G G' : SimpleGraph α}
+
+/--
+The diameter is the greatest distance between any two vertices, with the value `0` in case the
+distances are not bounded above.
+-/
+noncomputable def diam (G : SimpleGraph α) : ℕ :=
+  sSup {d | ∃ u v, d = G.dist u v}
+
+lemma diam_ne_zero_nonempty (h : G.diam ≠ 0) : Nonempty α := by
+  contrapose h
+  unfold diam
+  aesop
+
+private lemma not_bddAbove_diam_eq_zero (h : ¬BddAbove {d | ∃ u v, d = G.dist u v}) :
+    G.diam = 0 := by
+  apply Set.infinite_of_not_bddAbove at h
+  rw [diam, Set.Infinite.Nat.sSup_eq_zero h]
+
+lemma diam_exists [Nonempty α] : ∃ u v, G.dist u v = G.diam := by
+  let s := {d | ∃ u v, d = G.dist u v}
+  let u := Classical.arbitrary α
+  by_cases h : BddAbove s
+  · have : s.Nonempty := ⟨0, u, u, dist_self.symm⟩
+    obtain ⟨u, v, huv⟩ := Nat.sSup_mem this h
+    use u, v, huv.symm
+  · exact not_bddAbove_diam_eq_zero h ▸ ⟨u, u, dist_self⟩
+
+lemma bddAbove_dist_le_diam (h : BddAbove {d | ∃ u v, d = G.dist u v}) :
+    ∀ u v, G.dist u v ≤ G.diam := by
+  rw [diam, Nat.sSup_def h]
+  aesop
+
+lemma diam_bot : (⊥ : SimpleGraph α).diam = 0 := by
+  unfold diam
+  by_cases h : Nonempty α
+  · have : {d | ∃ u v, d = (⊥ : SimpleGraph α).dist u v} = {0} :=
+      Set.ext_iff.mpr fun _ ↦ ⟨fun ⟨_, _, h⟩ ↦ dist_bot _ _ ▸ h,
+      fun h ↦ ⟨Classical.arbitrary α, Classical.arbitrary α, dist_bot _ _ ▸ h⟩⟩
+    exact this ▸ csSup_singleton _
+  · simp_all
+
+lemma diam_eq_zero : G.diam = 0 ↔ ¬BddAbove {d | ∃ u v, d = G.dist u v} ∨ G = ⊥ := by
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · by_cases h' : G = ⊥
+    · exact Or.inr h'
+    · have : ∃ u v, G.Adj u v := by
+        by_contra
+        have : G = emptyGraph α := by ext; simp_all
+        exact h' (emptyGraph_eq_bot _ ▸ this)
+      obtain ⟨u, v, huv⟩ := this
+      apply Or.inl
+      by_contra h
+      have := bddAbove_dist_le_diam h u v
+      simp_all [dist_eq_one_iff_adj.mpr huv]
+  · cases' h with h h
+    · exact not_bddAbove_diam_eq_zero h
+    · rw [h, diam_bot]
+
+lemma diam_le (h : G.diam ≠ 0) : ∀ u v, G.dist u v ≤ G.diam := by
+  intros
+  apply le_csSup
+  rw [ne_eq, diam_eq_zero] at h
+  repeat tauto
+
+lemma diam_le_subgraph_diam [Nonempty α] (hg: G.Connected) (hz : G.diam ≠ 0) (h : G ≤ G') :
+    G'.diam ≤ G.diam :=
+  have ⟨u, v, huv⟩ := G'.diam_exists
+  huv ▸ LE.le.trans (dist_le_subgraph_dist h (hg u v)) (G.diam_le hz u v)
+
+/--
+The extended diameter is the greatest distance between any two vertices, with the value `⊤` in
+case the distances are not bounded above.
+-/
+noncomputable def ediam (G : SimpleGraph α) : ℕ∞ :=
+  sSup {d | ∃ u v : α, d = G.dist u v}
```

Comment:
... and similarly here the e-diam of a disconnected graph will be the maximum of the e-diam of its components instead of being infinite.

I think the correct definition is
```suggestion
/--
The extended diameter is the greatest distance between any two vertices, with the value `⊤` in
case the distances are not bounded above.
-/
noncomputable def ediam (G : SimpleGraph α) : ℕ∞ := ⨆ u v, G.edist u v
```
where `G.dist u v = (G.edist u v).toNat`. Then
```
noncomputable def diam (G : SimpleGraph α) : ℕ∞ := G.ediam.toNat
```
will be the correct notion

Reply:
Yes, that would be great!

Reply:
Does this approach seem good? There is still a lot of work do be done (I've marked the PR as WIP), but I would like to check that I'm not building the API on wrong foundations. Also I would love to hear @b-mehta's opinion on this.

Reply:
Yeah I think Yaël's suggestion is the right one

Reply:
I didn't mean to imply that Yaël's isn't right, instead, I'm wondering if this type of proofs using the weird `ediam_def` are okay. Sorry for the ambiguous question!